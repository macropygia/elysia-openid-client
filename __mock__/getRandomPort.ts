export function getRandomPort() {
  const server = Bun.serve({
    fetch() {
      return new Response("Bun!");
    },
    port: 0,
  });
  const currentPort = server.port;
  server.stop();
  return currentPort;
}
