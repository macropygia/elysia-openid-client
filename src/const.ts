import type {
  OIDCClientAuthHookSettings,
  OIDCClientCookieSettings,
  OIDCClientSettings,
} from "@/types";
import { t } from "elysia";
import type { LoggerOptions } from "pino";

/**
 * Plugin default settings
 */
export const defaultSettings: OIDCClientSettings = {
  callbackCompletedPath: "/",
  logoutCompletedPath: "/logout",
  pathPrefix: "/auth",
  loginPath: "/login",
  callbackPath: "/callback",
  logoutPath: "/logout",
  userinfoPath: "/userinfo",
  refreshPath: "/refresh",
  introspectPath: "/introspect",
  revokePath: "/revoke",
  statusPath: "/status",
  claimsPath: "/claims",
  resourcePath: "/resource",
  loginExpiration: 60 * 10 * 1000, // 10 minutes
  refreshExpiration: 60 * 60 * 24 * 30 * 1000, // 30 days
};

/**
 * Plugin default cookie settings
 */
export const defaultCookieSettings: OIDCClientCookieSettings = {
  sessionIdName: "__Host-sid",
  httpOnly: true,
  secure: true,
  sameSite: "lax",
  path: "/",
};

/**
 * Auth hook settings
 */
export const defaultAuthHookSettings: Omit<
  OIDCClientAuthHookSettings,
  "loginRedirectUrl"
> = {
  // loginRedirectUrl: `${pathPrefix}${loginPath}`,
  disableRedirect: false,
  autoRefresh: true,
};

/**
 * Default pino options
 * - Do not add `never`
 */
export const defaultLoggerOptions: Record<string, LoggerOptions> = {
  default: {
    level: "trace",
    transport: {
      target: "pino-pretty",
      options: {
        sync: true,
      },
    },
  },
  development: {
    level: "trace",
    transport: {
      target: "pino-pretty",
      options: {
        sync: true,
      },
    },
  },
  production: {
    level: "info",
    transport: {
      target: "pino-pretty",
      options: {
        sync: true,
      },
    },
  },
};

export const sessionDataTypeBox = t.Object({
  sessionId: t.String(),
  sessionExpiresAt: t.Number(),
  idToken: t.String(),
  accessToken: t.String(),
  refreshToken: t.Optional(t.String()),
});
