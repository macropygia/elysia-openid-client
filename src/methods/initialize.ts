import type { BaseOidcClient } from "@/core/BaseOidcClient";
import { validateOptions } from "@/utils/validateOptions";
import { Issuer } from "openid-client";

export async function initialize(this: BaseOidcClient) {
  const { issuerUrl, logger } = this;

  logger?.trace("methods/initialize");

  // Discover IdP by discovery endpoint
  this.issuer = await Issuer.discover(issuerUrl);

  validateOptions(this);

  // Initialize RP
  this.client = new this.issuer.Client(this.clientMetadata);

  // Issuer identifier
  // - For multiple issuers
  // - Identical to iss claim in the ID token
  // Ref. https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
  this.clients[this.issuer.metadata.issuer] = this.client;

  this.initialized = true;
}
