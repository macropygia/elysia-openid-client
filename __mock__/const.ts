import { mock } from "bun:test";
import type { OidcClient } from "@/core/OidcClient";
import { defaultCookieSettings, defaultSettings } from "@/core/const";
import { consoleLogger } from "@/loggers/consoleLogger";
import type {
  OIDCClientActiveSession,
  OIDCClientOptions,
  OIDCClientSession,
} from "@/types";
import type { Cookie } from "elysia";
import type { IdTokenClaims } from "openid-client";

export type DeepPartial<T> = T extends object
  ? { [P in keyof T]?: DeepPartial<T[P]> }
  : T;

export const sqliteMockFile = "__mock__/mock.sqlite";
export const sqliteTempFile = "__mock__/temp.sqlite";
export const lokiMockFile = "__mock__/mock.db";
export const lokiTempFile = "__mock__/temp.db";
export const lowMockFile = "__mock__/mock.json";
export const lowTempFile = "__mock__/temp.json";
export const redisPort = 6379;
export const redisHost = "localhost";
export const rpPort = 57828;
export const opPort = 57829;

export const logger = consoleLogger();

export const mockLogger = {
  silent: mock(),
  trace: mock(),
  debug: mock(),
  info: mock(),
  warn: mock(),
  error: mock(),
  fatal: mock(),
};

export const mockBaseOptions = {
  baseUrl: `https://localhost:${rpPort}`,
  issuerUrl: `https://localhost:${opPort}`,
  clientMetadata: {
    client_id: "mock-client-id",
    client_secret: "mock-client-secret",
  },
  authParams: {},
  settings: defaultSettings,
  cookieSettings: {
    ...defaultCookieSettings,
    secure: false,
  },
} as OIDCClientOptions;

export const mockIdToken =
  "eyJhbGciOiJSUzI1NiIsImtpZCI6IjFlOWdkazcifQ.ewogImlzcyI6Imh0dHBzOi8vc2VydmVyLmV4YW1wbGUuY29tIiwKICAic3ViIjoiMjQ4Mjg5NzYxMDAxIiwKICAiYXVkIjoiczZCaGRSa3F0MyIsCiAgIm5vbmNlIjoibi0wUzZfV3pBMk1qIiwKICAiZXhwIjoxMzExMjgxOTcwLAogImlhdCI6MTMxMTI4MDE3MAp9.ggW8hZ1EuVLuxNuuIJKX_V8a_OMXzR0EHrAsNSbvOojeUFFrEalG9covNi1SDYmXKlgNuL59mVPeHspqI6J7sbo4-K3TSkhogU6JUgiy3OdDrfAav9yU5SCLZSxnW8NAkFguHiMeJfRugKyybartdLriPmIYkmI6VzwlHxGiuqlEkjFZjoiewpIdRHFuzkzGSmXEOpZGDaSIPuP7oGxjn_vSI7v7kpJIlBS-xloiFqXnudkjESaWCtkZO5LF1Wn01YklmTPQoh3Sxsm0Tp1QTwecRjyFz7exbcVwhQqXYTlvaUo-4KE5lKcxKVV_uCZuhSA2QS1cvI";

export const mockIdTokenClaims = {
  aud: "s6BhdRkqt3",
  exp: 1311281970,
  iat: 1311280170,
  iss: "https://server.example.com",
  nonce: "n-0S6_WzA2Mj",
  sub: "248289761001",
};

export const mockLoginSession: OIDCClientSession = {
  sessionId: "mock-session-id",
  codeVerifier: "mock-code-verifier",
  state: "mock-state",
  nonce: "mock-nonce",
  sessionExpiresAt: 5000000000000,
};

export const mockActiveSession: OIDCClientActiveSession = {
  sessionId: "mock-session-id",
  idToken: "mock-id-token",
  accessToken: "mock-access-token",
  refreshToken: "mock-refresh-token",
  sessionExpiresAt: 5000000000000,
};

export const mockActiveSessionWithoutRefreshToken: OIDCClientActiveSession = {
  sessionId: "mock-session-id",
  idToken: "mock-id-token",
  accessToken: "mock-access-token",
  sessionExpiresAt: 5000000000000,
};

export const mockActiveSessionWithRealIdToken: OIDCClientActiveSession = {
  sessionId: "mock-session-id",
  idToken: mockIdToken,
  accessToken: "mock-access-token",
  refreshToken: "mock-refresh-token",
  sessionExpiresAt: 5000000000000,
};

export const mockPostInit: RequestInit = {
  method: "POST",
  headers: {
    Cookie: `${defaultCookieSettings.sessionIdName}=mock-sid`,
  },
};

export const mockPostInitWithSid = (sid?: string): RequestInit => ({
  method: "POST",
  headers: {
    Cookie: `${defaultCookieSettings.sessionIdName}=${sid || "mock-sid"}`,
  },
});

export const mockGetInit = (sid?: string): RequestInit => ({
  headers: {
    Cookie: `${defaultCookieSettings.sessionIdName}=${sid || "mock-sid"}`,
  },
});

export const mockClaims = {
  exp: 5000000000000 / 1000,
  iss: `http://localhost:${opPort}`,
  sub: "mock-sub",
} as IdTokenClaims;

export const mockStatus = {
  sessionExpiresAt: mockActiveSession.sessionExpiresAt,
  hasRefreshToken: !!mockActiveSession.refreshToken,
  isExpired: mockClaims.exp * 1000 < Date.now(),
  expiresAt: mockClaims.exp * 1000,
  issuerUrl: mockClaims.iss,
  sub: mockClaims.sub,
};

export const mockBaseClient = {
  ...mockBaseOptions,
  create: mock(),
  initialize: mock(),
  validateOptions: mock(),
  createSession: mock(),
  updateSession: mock().mockReturnValue(mockActiveSession),
  fetchSession: mock(),
  deleteSession: mock(),
  getSessionIdCookieType: mock(),
  getCookieDefinition: mock(),
  getAuthHook: mock(),
  getEndpoints: mock(),
  logger: undefined,
} as DeepPartial<OidcClient> as OidcClient;

export const mockCookie = {
  [defaultCookieSettings.sessionIdName]: {
    value: mockActiveSession.sessionId,
    remove: mock(),
    update: mock(),
  },
} as unknown as Record<string, Cookie<string>>;
