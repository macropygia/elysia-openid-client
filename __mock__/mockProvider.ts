import { logger, opPort } from "@/__mock__/const";
import Elysia from "elysia";
// biome-ignore lint/style/noNamespaceImport: <explanation>
import * as jose from "jose";
import loki from "lokijs";
import {
  type AuthorizationParameters,
  type IdTokenClaims,
  type TokenSetParameters,
  generators,
} from "openid-client";
import { mockIssuerMetadata } from "./issuerMetadata";

export const mockProvider = async (port: number) => {
  // JWT
  const keys = [];
  const { privateKey } = await jose.generateKeyPair("RS256", {
    extractable: true,
  });
  const { n, e, kty } = await jose.exportJWK(privateKey);
  const alg = "RS256";
  const kid = generators.random();
  const use = "sig";
  keys.push({ kty, alg, kid, use, e, n });
  const jwks = { keys };

  // DB
  const db = new loki("in-memory.db");
  interface Connection extends AuthorizationParameters {
    code?: string;
  }
  const connections = db.addCollection<Connection>("connections", {
    indices: [],
  });

  return new Elysia()
    .get("/oauth2/authorize", ({ request, set }) => {
      logger?.trace("/oauth2/authorize");
      const url = new URL(request.url);
      const state = url.searchParams.get("state") as string;
      const redirect_uri = url.searchParams.get("redirect_uri") as string;
      const code = generators.random();
      const connection: Connection = { code };
      for (const [key, value] of url.searchParams) {
        connection[key] = value;
      }
      connections.insert(connection);

      const params = new URLSearchParams({
        code,
        state,
      });
      set.status = 302;
      set.redirect = `${redirect_uri}?${params}`;
    })
    .post("/oauth2/token", async ({ body, set }) => {
      logger?.trace("/oauth2/token");
      const {
        // grant_type,
        code,
        // redirect_uri,
        code_verifier,
        client_id,
        // client_secret,
      } = body as unknown as Record<string, string>;
      const connection = connections.findOne({
        code: { $eq: code },
      }) as Connection;
      const { nonce, scope, code_challenge } = connection;

      if (code_challenge === generators.codeChallenge(code_verifier)) {
        logger?.trace("PKCE Code Verified");
      }

      const claims: IdTokenClaims = {
        iss: `http://localhost:${opPort}`,
        sub: "elysia-openid-client",
        aud: client_id,
        exp: Math.floor(Date.now() / 1000) + 3600,
        iat: Math.floor(Date.now() / 1000),
        nonce,
      };
      const id_token = await new jose.SignJWT(claims)
        .setProtectedHeader({ alg })
        .sign(privateKey);
      const tokenSet: TokenSetParameters = {
        access_token: "mock-access-token",
        token_type: "Bearer",
        id_token,
        refresh_token: "mock-refresh-token",
        scope,
        expires_at: Math.floor(Date.now() / 1000) + 60 * 60, // 1h
        session_state: "string",
        sub: "elysia-openid-client",
      };

      set.headers["Content-type"] = "application/json";
      return tokenSet;
    })
    .all("/oauth2/userinfo", () => {
      logger?.trace("/oauth2/userinfo");
      return { endpoint: "userinfo" };
    })
    .get("/oauth2/clients", () => {
      logger?.trace("/oauth2/clients");
      return "clients";
    })
    .get("/oauth2/keys", ({ set }) => {
      set.headers["Content-Type"] = "application/json";
      return jwks;
    })
    .post("/oauth2/introspect", () => {
      logger?.trace("/oauth2/introspect");
      return { endpoint: "introspect" };
    })
    .post("/oauth2/revoke", () => {
      logger?.trace("/oauth2/revoke");
      return "revoke";
    })
    .all("/oauth2/logout", ({ request }) => {
      logger?.trace(request.url);
      return "logout";
    })
    .get("/oauth2/device/authorize", ({ request }) => {
      logger?.trace(request.url);
      return "/device/authorize";
    })
    .get("/.well-known/openid-configuration", ({ set }) => {
      set.headers["Content-Type"] = "application/json";
      return mockIssuerMetadata;
    })
    .listen(port);
};
