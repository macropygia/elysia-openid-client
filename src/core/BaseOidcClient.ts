import { defaultCookieSettings, defaultSettings } from "@/const";
import { SQLiteAdapter } from "@/dataAdapters/SQLiteAdapter";
import { consoleLogger } from "@/loggers/consoleLogger";
import { initialize } from "@/methods/initialize";
import type {
  OIDCClientCookieSettings,
  OIDCClientDataAdapter,
  OIDCClientLogger,
  OIDCClientOptions,
  OIDCClientPaths,
  OIDCClientSettings,
} from "@/types";
import type {
  AuthorizationParameters,
  BaseClient,
  ClientMetadata,
  Issuer,
  IssuerMetadata,
} from "openid-client";

/**
 * OidcClient without methods for testing
 */
export class BaseOidcClient {
  /** Client Base URL */
  baseUrl: string;
  /** OIDC IdP URL */
  issuerUrl: string;
  /** OIDC Client metadata */
  clientMetadata: ClientMetadata;
  /** OIDC Authorization parameters */
  authParams: AuthorizationParameters;
  /** Plugin settings */
  settings: OIDCClientSettings;
  /** Cookie settings */
  cookieSettings: OIDCClientCookieSettings;
  /** OIDC Issuer (Initialize at create()) */
  issuer!: Issuer<BaseClient>;
  /** OIDC Issuer metadata (Initialize at create()) */
  issuerMetadata!: IssuerMetadata;
  /** OIDC Client (Initialize at create()) */
  client!: BaseClient;
  /** Plugin database */
  sessions: OIDCClientDataAdapter;
  /** Logger */
  logger: OIDCClientLogger | undefined;
  /** Initialized */
  initialized = false;

  /**
   * Constructor
   * @protected
   * @param options Options
   */
  protected constructor(options: OIDCClientOptions) {
    const {
      baseUrl,
      issuerUrl,
      clientMetadata,
      authParams,
      dataAdapter,
      settings,
      cookieSettings,
      logger,
    } = options;

    if (!issuerUrl) {
      throw new Error("issuerUrl is required");
    }

    this.baseUrl = baseUrl;
    this.issuerUrl = issuerUrl;
    this.clientMetadata = clientMetadata;
    this.authParams = authParams || {};
    this.settings = {
      ...defaultSettings,
      ...settings,
    };
    this.cookieSettings = {
      ...defaultCookieSettings,
      ...cookieSettings,
    };

    this.sessions = dataAdapter || new SQLiteAdapter();

    this.logger =
      logger === null ? undefined : logger ? logger : consoleLogger("info");
  }

  /**
   * Create BaseOidcClient instance
   * @static
   * @param options
   * @returns BaseOidcClient instance
   */
  static async create(options: OIDCClientOptions) {
    const instance = new BaseOidcClient(options);
    await instance.initialize();
    return instance;
  }

  /**
   * Async part of constructor
   * @protected
   */
  protected initialize = async () => initialize.call(this);

  /**
   * Get list of endpoint paths
   * @public
   * @returns Record<string, string>
   */
  public getPaths = (): OIDCClientPaths => {
    const {
      settings,
      settings: { pathPrefix },
    } = this;
    return {
      callback: `${pathPrefix}${settings.callbackPath}`,
      claims: `${pathPrefix}${settings.claimsPath}`,
      introspect: `${pathPrefix}${settings.introspectPath}`,
      login: `${pathPrefix}${settings.loginPath}`,
      logout: `${pathPrefix}${settings.logoutPath}`,
      refresh: `${pathPrefix}${settings.refreshPath}`,
      resource: `${pathPrefix}${settings.resourcePath}`,
      revoke: `${pathPrefix}${settings.revokePath}`,
      status: `${pathPrefix}${settings.statusPath}`,
      userinfo: `${pathPrefix}${settings.userinfoPath}`,
    };
  };
}
