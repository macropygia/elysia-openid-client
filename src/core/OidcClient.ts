import type * as http from "node:http";
import { introspect } from "@/methods/client/introspect";
import { refresh } from "@/methods/client/refresh";
import { resource } from "@/methods/client/resource";
import { revoke } from "@/methods/client/revoke";
import { userinfo } from "@/methods/client/userinfo";
import { createAuthHook } from "@/methods/createAuthHook";
import { createEndpoints } from "@/methods/createEndpoints";
import { createSession } from "@/methods/session/createSession";
import { deleteSession } from "@/methods/session/deleteSession";
import { fetchSession } from "@/methods/session/fetchSession";
import { updateSession } from "@/methods/session/updateSession";
import type {
  OIDCClientActiveSession,
  OIDCClientMethodArgs,
  OIDCClientOptions,
} from "@/types";
import { t } from "elysia";
import type {
  IdTokenClaims,
  IntrospectionResponse,
  TokenSet,
  UnknownObject,
  UserinfoResponse,
} from "openid-client";
import { BaseOidcClient } from "./BaseOidcClient";

/**
 * OpenID Connect client plugin for ElysiaJS
 * - Usage:
 *   - `const client = await BaseOidcClient.factory(options);`
 *   - `const endpoints = client.createEndpoints();`
 *   - `const hook = client.createAuthHook();`
 */
export class OidcClient extends BaseOidcClient {
  protected constructor(options: OIDCClientOptions) {
    super(options);

    this.refresh = refresh.bind(this);
    this.revoke = revoke.bind(this);
    this.userinfo = userinfo.bind(this);
    this.introspect = introspect.bind(this);
    this.resource = resource.bind(this);

    this.createSession = createSession.bind(this);
    this.updateSession = updateSession.bind(this);
    this.fetchSession = fetchSession.bind(this);
    this.deleteSession = deleteSession.bind(this);
  }

  /**
   * Create OidcClient instance
   * @param options
   * @static
   * @returns OidcClient instance
   */
  static async factory(options: OIDCClientOptions) {
    const instance = new OidcClient(options);
    await instance.initialize();
    return instance;
  }

  /** Use refresh endpoint directly */
  public refresh: (args: OIDCClientMethodArgs) => Promise<IdTokenClaims | null>;

  /** Use revoke endpoint directly */
  public revoke: (args: OIDCClientMethodArgs) => Promise<undefined | null>;

  /** Use userinfo endpoint directly */
  public userinfo: (
    args: OIDCClientMethodArgs,
  ) => Promise<UserinfoResponse<UnknownObject, UnknownObject> | null>;

  /** Use introspect endpoint directly */
  public introspect: (
    args: OIDCClientMethodArgs,
  ) => Promise<IntrospectionResponse | null>;

  /** Use resource endpoint directly */
  public resource: (
    args: OIDCClientMethodArgs & { resourceUrl: string },
  ) => Promise<({ body?: Buffer } & http.IncomingMessage) | null>;

  /**
   * Create session and insert to DB
   * @public
   * @returns [sessionId, authorizationUrl]
   */
  public createSession: () => Promise<[string, string]>;

  /**
   * Update session in DB
   * @public
   * @param sessionId Session ID
   * @param tokenSet TokenSet
   */
  public updateSession: (
    sessionId: string,
    tokenSet: TokenSet,
  ) => Promise<OIDCClientActiveSession | null>;

  /**
   * Find and validate session from cookie and DB
   * @public
   * @param sessionId Sessison ID
   * @returns Session data or false
   */
  public fetchSession: (
    sessionId: string | undefined,
  ) => Promise<OIDCClientActiveSession | null>;

  /**
   * Delete session from DB
   * @protected
   * @param sessionId Session ID
   */
  public deleteSession: (sessionId: string) => Promise<void>;

  /**
   * Cookie definition for ElysiaJS
   * @public
   */
  public get cookieTypeBox() {
    return t.Cookie({
      [this.cookieSettings.sessionIdName]: t.Optional(t.String()),
    });
  }

  /**
   * Get onBeforeHandle for restricted endpoints
   * @public
   * @returns ElysiaJS Plugin
   */
  public get authHook() {
    return createAuthHook.call(this);
  }

  /**
   * OpenID Connect client plugin for ElysiaJS
   * @public
   * @returns ElysiaJS Plugin
   */
  public get endpoints() {
    return createEndpoints.call(this);
  }
}
