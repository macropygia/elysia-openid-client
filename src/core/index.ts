import { BaseOidcClient } from "./BaseOidcClient";
import { OidcClient } from "./OidcClient";

export { BaseOidcClient, OidcClient };
