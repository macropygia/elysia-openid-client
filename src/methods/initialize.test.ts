import { afterAll, describe, expect, test } from "bun:test";
import { BaseOidcClient } from "@/core/BaseOidcClient";
import { mockBaseOptions } from "@mock/const";
import { getRandomPort } from "@mock/getRandomPort";
import { mockProvider } from "@mock/mockProvider";
import { initialize } from "./initialize";

describe("Unit/methods/initialize", async () => {
  const opPort = getRandomPort();

  const op = await mockProvider(opPort);

  test("Default", async () => {
    const mockOptions = structuredClone(mockBaseOptions);
    const mockClinet = await BaseOidcClient.factory({
      ...mockOptions,
      issuerUrl: `http://localhost:${opPort}`,
    });
    expect(initialize.call(mockClinet)).resolves.toBeUndefined();
    expect(mockClinet.issuer).toMatchSnapshot("issuer");
    // expect(mockClinet.issuerMetadata).toMatchSnapshot("issuerMetadata");
    expect(mockClinet.client).toMatchSnapshot("client");
    expect(mockClinet.initialized).toBeTrue();
  });

  afterAll(async () => {
    await op.stop();
  });
});
