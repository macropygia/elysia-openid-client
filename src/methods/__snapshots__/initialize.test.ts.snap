// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit/methods/initialize Default: issuer 1`] = `
Issuer {
  "Client": [class Client],
  "FAPI1Client": [class FAPI1Client],
  "authorization_endpoint": [native code],
  "claim_types_supported": [native code],
  "claims_parameter_supported": [native code],
  "claims_supported": [native code],
  "code_challenge_methods_supported": [native code],
  "device_authorization_endpoint": [native code],
  "dpop_signing_alg_values_supported": [native code],
  "end_session_endpoint": [native code],
  "grant_types_supported": [native code],
  "id_token_signing_alg_values_supported": [native code],
  "introspection_endpoint": [native code],
  "introspection_endpoint_auth_methods_supported": [native code],
  "issuer": [native code],
  "jwks_uri": [native code],
  "registration_endpoint": [native code],
  "request_object_signing_alg_values_supported": [native code],
  "request_parameter_supported": [native code],
  "request_uri_parameter_supported": [native code],
  "require_request_uri_registration": [native code],
  "response_modes_supported": [native code],
  "response_types_supported": [native code],
  "revocation_endpoint": [native code],
  "revocation_endpoint_auth_methods_supported": [native code],
  "scopes_supported": [native code],
  "subject_types_supported": [native code],
  "token_endpoint": [native code],
  "token_endpoint_auth_methods_supported": [native code],
  "userinfo_endpoint": [native code],
}
`;

exports[`Unit/methods/initialize Default: issuerMetadata 1`] = `
{
  "authorization_endpoint": "http://localhost:57829/oauth2/authorize",
  "claim_types_supported": [
    "normal",
  ],
  "claims_parameter_supported": false,
  "claims_supported": [
    "iss",
    "ver",
    "sub",
    "aud",
    "iat",
    "exp",
    "jti",
    "auth_time",
    "amr",
    "idp",
    "nonce",
    "name",
    "nickname",
    "preferred_username",
    "given_name",
    "middle_name",
    "family_name",
    "email",
    "email_verified",
    "profile",
    "zoneinfo",
    "locale",
    "address",
    "phone_number",
    "picture",
    "website",
    "gender",
    "birthdate",
    "updated_at",
    "at_hash",
    "c_hash",
  ],
  "code_challenge_methods_supported": [
    "S256",
  ],
  "device_authorization_endpoint": "http://localhost:57829/oauth2/device/authorize",
  "dpop_signing_alg_values_supported": [
    "RS256",
    "RS384",
    "RS512",
    "ES256",
    "ES384",
    "ES512",
  ],
  "end_session_endpoint": "http://localhost:57829/oauth2/logout",
  "grant_types_supported": [
    "authorization_code",
    "refresh_token",
  ],
  "id_token_signing_alg_values_supported": [
    "RS256",
  ],
  "introspection_endpoint": "http://localhost:57829/oauth2/introspect",
  "introspection_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post",
    "client_secret_jwt",
    "private_key_jwt",
    "none",
  ],
  "issuer": "http://localhost:57829",
  "jwks_uri": "http://localhost:57829/oauth2/keys",
  "registration_endpoint": "http://localhost:57829/oauth2/clients",
  "request_object_signing_alg_values_supported": [
    "HS256",
    "HS384",
    "HS512",
    "RS256",
    "RS384",
    "RS512",
    "ES256",
    "ES384",
    "ES512",
  ],
  "request_parameter_supported": true,
  "request_uri_parameter_supported": true,
  "require_request_uri_registration": false,
  "response_modes_supported": [
    "query",
  ],
  "response_types_supported": [
    "code",
  ],
  "revocation_endpoint": "http://localhost:57829/oauth2/revoke",
  "revocation_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post",
    "client_secret_jwt",
    "private_key_jwt",
    "none",
  ],
  "scopes_supported": [
    "openid",
    "email",
    "profile",
    "address",
    "phone",
    "offline_access",
    "groups",
  ],
  "subject_types_supported": [
    "public",
  ],
  "token_endpoint": "http://localhost:57829/oauth2/token",
  "token_endpoint_auth_methods_supported": [
    "client_secret_post",
    "none",
  ],
  "userinfo_endpoint": "http://localhost:57829/oauth2/userinfo",
}
`;

exports[`Unit/methods/initialize Default: client 1`] = `
Client {
  "authorization_signed_response_alg": [native code],
  "client_id": [native code],
  "client_secret": [native code],
  "grant_types": [native code],
  "id_token_signed_response_alg": [native code],
  "introspection_endpoint_auth_method": [native code],
  "redirect_uris": [native code],
  "response_types": [native code],
  "revocation_endpoint_auth_method": [native code],
  "token_endpoint_auth_method": [native code],
}
`;

exports[`Unit/methods/initialize Default: clients 1`] = `
Client {
  "authorization_signed_response_alg": [native code],
  "client_id": [native code],
  "client_secret": [native code],
  "grant_types": [native code],
  "id_token_signed_response_alg": [native code],
  "introspection_endpoint_auth_method": [native code],
  "redirect_uris": [native code],
  "response_types": [native code],
  "revocation_endpoint_auth_method": [native code],
  "token_endpoint_auth_method": [native code],
}
`;
